[TSK-819] [TSK-820] ui-kit:Rich-text-element

jatinrise opened the pull request:
## Description

<!-- Provide a clear and concise description of the changes made in this pull request for UI-Kit components. -->

- Component name: `RichText-editor Edit and view mode` 
### This PR introduces a fully-featured Rich Text Editor and a Read-Only Renderer using Slate.js, along with Storybook integration for better UI development and testing.

- **Keyboard Shortcuts:**
- Ctrl + B → Bold
- Ctrl + I → Italic
- Ctrl + U → Underline
- Ctrl + Z/Y → Undo/Redo
   
<img width="735" alt="Screenshot 2025-03-24 at 22 10 17" src="https://github.com/user-attachments/assets/13232d1c-8a17-4354-b051-98b9cbe4d47c" />






## Checklist

- [x] I have run the lint, build, and test commands locally, as explained in the README, before creating this PR to ensure the code is clean and functional.
- [x] My changes align with the **core models** that have been discussed and agreed upon.
- [x] The component design has been compared to the **Figma design** to ensure consistency.
- [x] The **component is fully responsive** and behaves as expected across different screen sizes.
- [x] The component **supports keyboard web accessibility** (if applicable).
- [x] I have written **unit tests** or **integration tests** to cover the component's functionality above the coverage threshold.
- [x] I have updated the **documentation** for the component, including usage examples and guidelines.
- [x] I have **commented** my code, particularly in complex or non-obvious areas.
- [x] The **component is properly exported** for use in other parts of the application.

---

## Related Issues

<!-- Link any related issue or task. For example: -->
Fixes #105 

Hey @alebg, check out my PR. I have made all the changes we discussed today.

alebg commented on packages/ui-kit/lib/utils/constants.tsx:
```diff
@@ -0,0 +1,135 @@
+    { id: RichTextAction.Bold, icon: <Bold  size={24}/>, label: "Bold(Ctrl+B)"},
+    { id: RichTextAction.Italics, icon: <Italic  size={24}/>, label: "Italics(Ctrl+I)"},
+    { id: RichTextAction.Underline, icon: <Underline size={24} />, label: "Underline(Ctrl+U)" },
+    {id: RichTextAction.link, icon: <Link size={24} />, label: "Link"},
```
@jatinrise please remove this Link. It makes the option appear twice:

![image](https://github.com/user-attachments/assets/2a7e8838-2f08-4106-8987-8485c7cb89db)

But this one doesn't work (the first one, next to the Unlink, does though)

alebg commented on packages/ui-kit/lib/components/rich-text-editor/editor.tsx:
```diff
@@ -0,0 +1,173 @@
+
+    return (
+      <div className="w-full text-text-primary">
+        <Slate editor={editor} initialValue={initialValue} onChange={(value) => onChange(value)}>
```
@jatinrise as discussed yesterday, could you please serialize to a string the value that goes out of the editor?

We don't want to send out slateJS objects, but something that the backend can handle directly (strings)

alebg replied:
It would just be something like `onChange(serialize(value))` here, assuming `serialize` is the name of your function

jatinrise replied:
@alebg We can't do this in the onChange function, but we can serialize the array when clicking a button and api will call.

alebg replied:
@jatinrise Yes I see the problem, indeed it's not that simple.

Ok so on further discussion with Mayank, let us try the following. So, let's leave the line I mentioned as is, as well as the onChange logic. Then, please pass a callback to this component, called `onLoseFocus`, which will receive a string as an input. Please implement some logic to have this callback executed once the user is not focused on the editor component anymore, and pass to this callback the serialized value.

The idea is that then, in the app (page), we will save to the backend everytime the user focuses out of the editor.

Please also adapt the stories such that you pass a console.log as this callback, so I can quickly grab the serialized values from the console and test your component more quickly.

alebg reviewed (changes requested):
@jatinrise great job, it's looking very good. Please address my in-line comments. Apart from that:

- [x] Please change the name of the directory from `rich-text-editor` to `rich-text-element`, as it contains more than just the editor
- [ ] Please be consistent with the naming of the components. So, `RichTextEditor` is good, but please change `SlateRenderer` to its counterpart, `RichTextRenderer`. And the file name of the latter, to keep it less confusing, from `view-mode.tsx` to `renderer.tsx`. So we know quickly where to look
- [ ] In my opinion, the editor should shrink horizontally with the screen. I don't remember if there was a reason why this isn't happening?
- [ ] Let's also wait for @carlapbg 's review

jatinrise commented:
> * RichTextRenderer

@alebg from my point of view scrollable toolbar is looking more suitable. because there are lot of function is there itovvertake the editor height

jatinrise commented:
@alebg, all the above inline comments have been incorporated. Can you please verify them?

carlapbg reviewed (changes requested):
Hi, in general looks so good, thank you! There is just something to change:

- Can you add the resize handle in the corner?
![Screenshot 2025-03-25 at 15 01 22](https://github.com/user-attachments/assets/9f652c40-a425-4900-ad41-a459771b9fab)

- Can you move the "start typing..." text a bit lower? Right where typing begins

<img width="409" alt="Screenshot 2025-03-25 at 15 08 03" src="https://github.com/user-attachments/assets/b1193186-209e-4a8b-bb69-fd49e991d139" />

- Why does the scrollbar appear at the bottom when nothing has been typed yet? When I start typing, it disappears. Can we remove it?

<img width="1131" alt="Screenshot 2025-03-25 at 15 11 26" src="https://github.com/user-attachments/assets/623d6e2d-8b01-4c3a-9087-06338365b6a2" />

- When the text is very long, the text box expands slightly to the right up to a certain point before the text starts wrapping to the next line. It would be better if this didn’t happen. The box should only grow vertically based on the text length, not to the right.

jatinrise commented:
@carlapbg Thank you for the review,i have done the inline comment changes,can you please verify.

carlapbg commented:
Thanks for the changes! Just a few things:
- The resize handle should only increase the box size in length. Right now, I can shrink it without any limits in both width and length

<img width="278" alt="Screenshot 2025-03-25 at 18 08 57" src="https://github.com/user-attachments/assets/ad088ab1-ac96-49ad-949f-3aa09e76b6d3" />
<img width="1020" alt="Screenshot 2025-03-25 at 18 09 06" src="https://github.com/user-attachments/assets/6eb3da75-2f2d-42ae-b353-089273de88c4" />


- As the screen gets smaller, the box should also shrink. Right now, on smaller screens, I have to scroll to see the full box.

<img width="405" alt="Screenshot 2025-03-25 at 18 12 29" src="https://github.com/user-attachments/assets/13107108-da06-4558-83d4-0cb6acfd80cd" />

- Normally, on Mac computers, the keyboard shortcuts for editing the text (bold, underline, italic, undo/redo) use the “Command” key instead of “Control”. Could you add that? That is:  Cmd + B → Bold Cmd + I → Italic Cmd + U → Underline Cmd + Z/Y → Undo/Redo

jatinrise commented:
@carlapbg @alebg Thank you for the review! All the changes below have been incorporated:

- **Keyboard accessibility**: Added support for Mac users (e.g., Cmd+B, Cmd+I, etc.).
- 
- **Responsive toolbar**: The toolbar now automatically shrinks when the screen width decreases.
- 
- **Focus handling**: The onFocusLose function is now passed as a prop. Additionally, I have created a Storybook entry that returns serialized output on onBlur.
- 
- **Resizing improvements**: The editor can now only be resized vertically, and I have also fixed its minimum width.

alebg commented on packages/ui-kit/lib/components/rich-text-element/editor.tsx:
```diff
@@ -128,45 +129,55 @@ export const RichTextEditor: React.FC<RichTextEditorProps> = React.memo(
+    const serialized = serialize(editor.children);
+    onLoseFocus(serialized);
+    console.log('serialized',serialized);
+   }
```
@jatinrise thanks a lot. Please don't hard-code a console.log here. What I meant is that you should be able to pass (e.g., in the story), a callback that just does console.log

jatinrise commented:
@alebg Thanks for the review.can you please verify now

alebg commented on packages/ui-kit/stories/rich-text-editor.stories.tsx:
```diff
@@ -39,7 +40,8 @@ export const Basic: Story = {
     placeholder: 'Start typing...',
     initialValue: defaultInitialValue,
-    onChange: (value) => console.log('Content changed:', value),
+    onChange: (value) => console.log('Content changed:'),
```
@jatinrise please take out these onChange callbacks, they pollute the console in an inconsistent way. In the story, lets keep only the onLoseFocus callbacks that console.log, those are working nicely

alebg reviewed (changes requested):
@jatinrise please address my two inline comments. Additionally, could you please add a space between the format name and its keyboard shortcut? E.g., on hover over the "B", it should say "Bold (Ctrl+B)", instead of "Bold(Ctrl+B)"

Apart from that, really great job! It's looking very good, and your onLoseFocus implementation works nicely.

alebg commented:
> @alebg Thanks for the review.can you please verify now.

@jatinrise Thanks, I asked for a couple more things

alebg commented on packages/ui-kit/stories/rich-text-editor.stories.tsx:
```diff
@@ -40,7 +40,7 @@ export const Basic: Story = {
     placeholder: 'Start typing...',
     initialValue: defaultInitialValue,
-    onChange: (value) => console.log('Content changed:'),
+    onChange: (value) => null,
```
@jatinrise Is it mandatory to pass this onChange callback always? I don't see why would we need it if we already have the onLoseFocus callback

jatinrise replied:
@carlapbg  I think this is used to solve a specific use case. Suppose we need to show both the editor and the renderer at the same time—in the assignment section for the coach, we have this use case.

alebg commented on packages/ui-kit/lib/utils/constants.tsx:
```diff
@@ -44,9 +44,9 @@ import {
-    { id: RichTextAction.Bold, icon: <Bold  size={24}/>, label: "Bold(Ctrl+B)"},
-    { id: RichTextAction.Italics, icon: <Italic  size={24}/>, label: "Italics(Ctrl+I)"},
-    { id: RichTextAction.Underline, icon: <Underline size={24} />, label: "Underline(Ctrl+U)" },
+    { id: RichTextAction.Bold, icon: <Bold  size={24}/>, label: "Bold (Ctrl+B)"},
```
@jatinrise Thanks a lot! Little detail I just noticed, sorry, is that these labels should be localized as well ("Bold", "Italics", etc.)

jatinrise replied:
@alebg Thank you for reminding me.

alebg commented on packages/ui-kit/lib/components/rich-text-element/types.ts:
```diff
@@ -1,15 +1,16 @@
   initialValue: Descendant[] | undefined | string;
-  onChange: (value: Descendant[]) => void;
+  onChange?: (value: Descendant[]) => void;
   onLoseFocus?: (value:string) => void;
```
@jatinrise could you make this (`onLoseFocus`) non-optional? So when using it we know what we should pass to get the expected behaviour. If both are optional, then we can have a case in which nothing will happen from the editor outwards.

This is my last remark. After this, whenever @carlapbg validates we're ready to merge

alebg commented:
> @alebg Thank you! Made the onLoseFocus prop optional.

I believe you mean mandatory :). Thanks a lot, and again, great job!

carlapbg reviewed (approved):
Looks good now! Thank you!
